src/common/mem.c

```diff
@@ -1573,4 +1573,473 @@ umem_tx_publish(struct umem_instance *umm, struct umem_rsrvd_act *rsrvd_act)
 	rsrvd_act->rs_actv_at = 0;
 	return rc;
 }
+
+int
+umem_cache_alloc(struct umem_store *store, uint64_t max_mapped)
+{
+	struct umem_cache *cache;
+	uint64_t           num_pages;
+	int                rc = 0;
+	int                idx;
+
+	D_ASSERT(store != NULL);
+
+	num_pages = (store->stor_size + UMEM_CACHE_PAGE_SZ - 1) >> UMEM_CACHE_PAGE_SZ_SHIFT;
+
+	if (max_mapped != 0) {
+		D_ERROR("Setting max_mapped is unsupported at present\n");
+		return -DER_NOTSUPPORTED;
+	}
+
+	D_ALLOC(cache, sizeof(*cache) + num_pages * sizeof(cache->ca_pages[0]));
+	if (cache == NULL)
+		D_GOTO(error, rc = -DER_NOMEM);
+
+	D_DEBUG(DB_IO,
+		"Allocated page cache for stor->stor_size=" DF_U64 ", " DF_U64 " pages at %p\n",
+		store->stor_size, num_pages, cache);
+
+	cache->ca_store      = store;
+	cache->ca_num_pages  = num_pages;
+	cache->ca_max_mapped = num_pages;
+
+	D_INIT_LIST_HEAD(&cache->ca_pgs_dirty);
+	D_INIT_LIST_HEAD(&cache->ca_pgs_copying);
+	D_INIT_LIST_HEAD(&cache->ca_pgs_lru);
+
+	for (idx = 0; idx < num_pages; idx++)
+		cache->ca_pages[idx].pg_id = idx;
+
+	store->cache = cache;
+
+	return 0;
+
+error:
+	D_FREE(cache);
+	return rc;
+}
+
+int
+umem_cache_free(struct umem_store *store)
+{
+	/** XXX: check reference counts? */
+	D_FREE(store->cache);
+	return 0;
+}
+
+int
+umem_cache_check(struct umem_store *store, uint64_t num_pages)
+{
+	struct umem_cache *cache = store->cache;
+
+	D_ASSERT(num_pages + cache->ca_mapped <= cache->ca_num_pages);
+
+	if (num_pages > cache->ca_max_mapped - cache->ca_mapped)
+		return num_pages - (cache->ca_max_mapped - cache->ca_mapped);
+
+	return 0;
+}
+
+int
+umem_cache_evict(struct umem_store *store, uint64_t num_pages)
+{
+	/** XXX: Not yet implemented */
+	return 0;
+}
+
+int
+umem_cache_map_range(struct umem_store *store, umem_off_t offset, void *start_addr,
+		     uint64_t num_pages)
+{
+	struct umem_cache *cache = store->cache;
+	struct umem_page *page;
+	struct umem_page *end_page;
+	uint64_t          current_addr = (uint64_t)start_addr;
+
+	if (store->cache == NULL)
+		return 0; /* TODO: When SMD is supported outside VOS, this will be an error */
+
+	page     = umem_cache_off2page(cache, offset);
+	end_page = page + num_pages;
+
+	D_ASSERTF(page->pg_id + num_pages <= cache->ca_num_pages,
+		  "pg_id=%d, num_pages=" DF_U64 ", cache pages=" DF_U64 "\n", page->pg_id,
+		  num_pages, cache->ca_num_pages);
+
+	while (page != end_page) {
+		D_ASSERT(page->pg_addr == NULL);
+		page->pg_addr = (void *)current_addr;
+		current_addr += UMEM_CACHE_PAGE_SZ;
+
+		d_list_add_tail(&page->pg_link, &cache->ca_pgs_lru);
+		page++;
+	}
+
+	cache->ca_mapped += num_pages;
+
+	return 0;
+}
+
+int
+umem_cache_pin(struct umem_store *store, umem_off_t addr, daos_size_t size)
+{
+	struct umem_cache *cache     = store->cache;
+	struct umem_page *page      = umem_cache_off2page(cache, addr);
+	struct umem_page *end_page  = umem_cache_off2page(cache, addr + size - 1) + 1;
```

Comment:
why do we need to "+1" at here? e.g. if addr=0, size=4K, we only pin one page not two right?